synapse/third_party_rules/access_rules.py

```diff
@@ -424,6 +432,22 @@ def _is_power_level_content_allowed(self, content, access_rule):
 
         return True
 
+    def _on_join_rule_change(self, event, rule):
+        """Check whether a join rule change is allowed. A join rule change is always
+        allowed unless the new join rule is "public" and the current access rule isn't
+        "restricted".
+
+        Args:
+            event (synapse.events.EventBase): The event to check.
+            rule (str): The name of the rule to apply.
+        Returns:
+            bool, True if the event can be allowed, False otherwise.
+        """
+        if event.content.get('join_rule') == JoinRules.PUBLIC:
+            return rule == ACCESS_RULE_RESTRICTED
```

Comment:
Shouldn't this be `!=`?

Reply:
Nope. We want to return `True` if the `m.room.join_rules` event can be allowed. The rules on this are documented in the docstring above:

> A join rule change is always allowed unless the new join rule is "public" and the current access rule isn't "restricted".

Reply:
Ok. Yes. That is slightly mind bending. That is equivalent to saying that publicly joinable rooms must be `restricted`. Got it.

Reply:
Yeah, the rationale is that external users should only be allowed access to rooms that are invite-only, i.e. they should always rely on an agent user for access to any room.

Reply:
I'll make the comment more explicit.